catch failed decryption and print a helpful message

GeoffWilliams opened the pull request:
## Background

This PR prints a helpful message on failed decryption by catching `OpenSSL::PKCS7::PKCS7Error` from OpenSSL.
## Issue

Addresses:  https://github.com/TomPoulton/hiera-eyaml/issues/143
Currently, the message is `PKCS7[Method: 112, Reason: -1]`
## Messages introduced by this PR

I tested output attempting to decrypt with the wrong key on both puppet:

```
Info: Retrieving pluginfacts
Info: Retrieving plugin
Info: Loading facts
Error: Could not retrieve catalog from remote server: Error 400 on SERVER: Hiera-eyaml decryption failed, check the encrypted data matches the key you are using.
Raw message from system: PKCS7[Method: 112, Reason: -1]
Warning: Not using cache on failed catalog
Error: Could not retrieve catalog; skipping run
```

And the command line:

```
[root@xmaster puppet]# hiera myvar
/opt/puppet/lib/ruby/gems/1.9.1/gems/hiera-eyaml-2.0.6/lib/hiera/backend/eyaml_backend.rb:89:in `rescue in decrypt': Hiera-eyaml decryption failed, check the encrypted data matches the key you are using. (Exception)
Raw message from system: bad decrypt
    from /opt/puppet/lib/ruby/gems/1.9.1/gems/hiera-eyaml-2.0.6/lib/hiera/backend/eyaml_backend.rb:81:in `decrypt'
 ...
```
## Additional Features
- The raw error message from OpenSSL is returned
- The backtrace is output using `debug()` and is visible if invoking `hiera` from the command line

vinzent commented on lib/hiera/backend/eyaml_backend.rb:
```diff
@@ -78,12 +78,19 @@ def debug(message)
+          rescue OpenSSL::PKCS7::PKCS7Error => e
+            debug("Caught exception: #{e.class}, #{e.message}\n"\
+                  "#{e.backtrace.join("\n")}")
+            raise Exception, "Hiera-eyaml decryption failed, check the "\
```
IMHO raising a general exception is not good practice. Why does it need to be `Exception`?

What about re-using `OpenSSL::PKCS7::PKCS7Error` with an adjusted message? Or just `raise "Message ..."` (which IMHO defaults to `RuntimeError`.

binford2k replied:
I'd vote for the `RuntimeError`

GeoffWilliams replied:
Your both right - I got lazy. Of course catching `Exception` is an anti-pattern. As for this PR - it got sidelined due to other work. let me see if I can update it

binford2k replied:
Ha! You're right. We should fix that... :)

bastelfreak commented:
Hi @GeoffWilliams, are you still interested in getting this merged? If so, could you please rebase and take a look at the comment from @vinzent?

GeoffWilliams commented:
Updated and rebased - I get the same cucumber failure on `master` vs this feature branch, so it doesn't look like anything new was broken by this:

```
geoff@computer:~/github/hiera-eyaml$ bundle exec cucumber -f progress
..................................................................................................................................................................................................................................F-............................DEPRECATION: Using `should` from rspec-expectations' old `:should` syntax without explicitly enabling the syntax is deprecated. Use the new `:expect` syntax or explicitly enable `:should` instead. Called from /home/geoff/github/hiera-eyaml/features/step_definitions/parser_steps.rb:27:in `block in <top (required)>'.
.....Decrypt_string: 
.........................Decrypt_string: 
......................................................Decrypt_string: 
Encrypt_string: 
Decrypt_string: 
..................................

(::) failed steps (::)

No such file or directory - ./supply_password.sh (Aruba::LaunchError)
features/encrypts.feature:24:in `When I run `./supply_password.sh eyaml encrypt -o string -p`'

Failing Scenarios:
cucumber features/encrypts.feature:23 # Scenario: encrypt a password

76 scenarios (1 failed, 75 passed)
374 steps (1 failed, 1 skipped, 372 passed)
0m40.018s
```
